array.h

```diff
@@ -808,6 +811,16 @@ T convert_dims(const U& u, std::index_sequence<Is...>) {
   return std::make_tuple(convert_dim<Is, T>(u)...);
 }
 
+constexpr index_t factorial(index_t x) {
+  return x == 1 ? 1 : x * factorial(x - 1);
+}
+
+// The errors that result from not satisfying this check are probably hell,
+// but it would be pretty tricky to check that all of [0, Rank) is in `Is...`
+template <size_t Rank, size_t... Is>
+using enable_if_permutation =
+    typename std::enable_if<product((Is + 2)...) == factorial(Rank + 1)>::type;
```

Comment:
enable_if_t ?

Reply:
I thought that was C++17, but it's C++14. I'll do it in a separate change though, because there are lots of std::enable_ifs that should be enable_if_t.

Reply:
https://stackoverflow.com/questions/19559808/constexpr-initialization-of-array-to-sort-contents

bubble / selection sort looks easy enough.

Then we can use is_same_as<T> to check the that integer sequences are equal.

Reply:
BTW, https://github.com/dsharlet/array/commit/dc06fa3936696007a655836f810681d47c090e66